src/ctap/storage.rs

```diff
@@ -315,171 +326,170 @@ impl PersistentStore {
             .find_one(&Key::GlobalSignatureCounter)
             .map_or(0, |(_, entry)| {
                 u32::from_ne_bytes(*array_ref!(entry.data, 0, 4))
-            })
+            }))
     }
 
-    pub fn incr_global_signature_counter(&mut self) {
+    pub fn incr_global_signature_counter(&mut self) -> Result<(), Ctap2StatusCode> {
         let mut buffer = [0; core::mem::size_of::<u32>()];
         match self.store.find_one(&Key::GlobalSignatureCounter) {
             None => {
                 buffer.copy_from_slice(&1u32.to_ne_bytes());
-                self.store
-                    .insert(StoreEntry {
-                        tag: GLOBAL_SIGNATURE_COUNTER,
-                        data: &buffer,
-                        sensitive: false,
-                    })
-                    .unwrap();
+                self.store.insert(StoreEntry {
+                    tag: GLOBAL_SIGNATURE_COUNTER,
+                    data: &buffer,
+                    sensitive: false,
+                })?;
             }
             Some((index, entry)) => {
                 let value = u32::from_ne_bytes(*array_ref!(entry.data, 0, 4));
                 // In hopes that servers handle the wrapping gracefully.
                 buffer.copy_from_slice(&value.wrapping_add(1).to_ne_bytes());
-                self.store
-                    .replace(
-                        index,
-                        StoreEntry {
-                            tag: GLOBAL_SIGNATURE_COUNTER,
-                            data: &buffer,
-                            sensitive: false,
-                        },
-                    )
-                    .unwrap();
+                self.store.replace(
+                    index,
+                    StoreEntry {
+                        tag: GLOBAL_SIGNATURE_COUNTER,
+                        data: &buffer,
+                        sensitive: false,
+                    },
+                )?;
             }
         }
+        Ok(())
     }
 
-    pub fn master_keys(&self) -> MasterKeys {
-        // We have as invariant that there is always exactly one MasterKeys entry in the store.
+    pub fn master_keys(&self) -> Result<MasterKeys, Ctap2StatusCode> {
         let (_, entry) = self.store.find_one(&Key::MasterKeys).unwrap();
-        let data = entry.data;
-        // And this entry is well formed: the encryption key followed by the hmac key.
-        let encryption = array_ref!(data, 0, 32);
-        let hmac = array_ref!(data, 32, 32);
-        MasterKeys { encryption, hmac }
+        if entry.data.len() != 64 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+        }
+        Ok(MasterKeys(entry.data.to_vec()))
     }
 
-    pub fn pin_hash(&self) -> Option<&[u8; PIN_AUTH_LENGTH]> {
-        self.store
-            .find_one(&Key::PinHash)
-            .map(|(_, entry)| array_ref!(entry.data, 0, PIN_AUTH_LENGTH))
+    pub fn pin_hash(&self) -> Result<Option<Vec<u8>>, Ctap2StatusCode> {
```

Comment:
Same here, why move from reference to arrays entirely, when `pin_hash` has a known fixed length?